README.md
# tokenish

![Code Coverage](https://img.shields.io/badge/Coverage-96%25-brightgreen.svg)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

A simple tool to fill pattern with tokens from file or directory.

- [Installation](#installation)
- [How to use](#how-to-use)
- [Example](#example)
- [Commands](#commands)
- [License](#license)
- [Tests](#tests)

## Installation

## How to use

```txt
usage: tokenish [-h] [-t TOKENS [TOKENS ...]] [-e ENCODING] [-o OUTPUT_DIRECTORY] [-om MAX_FILE_ROWS] pattern

Generate rows from pattern for each token combinations

positional arguments:
  pattern               text to fill with links or usernames/passwords

optional arguments:
  -h, --help            show this help message and exit
  -t TOKENS [TOKENS ...], --tokens TOKENS [TOKENS ...]
                        list of tokens file or directory path
  -e ENCODING, --encoding ENCODING
                        type of encoding to apply
  -o OUTPUT_DIRECTORY, --output-directory OUTPUT_DIRECTORY
                        directory where results will write, print is None
  -om MAX_FILE_ROWS, --max-file-rows MAX_FILE_ROWS
                        maximum number of rows per file, default 10 000
```

Available token paths:

- file: all lines in file are consider like tokens.
- directory: all lines of all files in directory are consider like tokens.

Each token paths should have an associated token in pattern.
For example, if two token paths are defined, pattern should contain &TOKEN_0& and &TOKEN_1&.

Available encodings:

- base64

The pattern part to encode must be in the following tag:

```txt
&ENC[to encode]ODE&
```

## Example

Let's consider the /tokens directory with the following tree structure:

```txt
.
├── passwords
│   └── passwords_1.txt
└── usernames
    ├── usernames_1.txt
    └── usernames_2.txt
```

Content of usernames_1.txt:

```txt
admin
root
user
```

Content of usernames_2.txt:

```txt
user1
user2
```

Content of passwords_1.txt:

```txt
1234
6712
```

The command:

```sh
tokenish "Authorization: Basic &TOKEN_0&:&TOKEN_1&" -t usernames/ passwords/passwords_1.txt
```

Print:

```txt
Authorization: Basic admin:1234
Authorization: Basic admin:6712
Authorization: Basic root:1234
Authorization: Basic root:6712
Authorization: Basic user:1234
Authorization: Basic user:6712
Authorization: Basic user1:1234
Authorization: Basic user1:6712
Authorization: Basic user2:1234
Authorization: Basic user2:6712
```

You can add an encoding like:

```sh
tokenish "Authorization: Basic &ENC[&TOKEN_0&:&TOKEN_1&]ODE&" -t usernames/ passwords/passwords_1.txt -e base64
```
Print:

```txt
Authorization: Basic YWRtaW46MTIzNA==
Authorization: Basic YWRtaW46NjcxMg==
Authorization: Basic cm9vdDoxMjM0
Authorization: Basic cm9vdDo2NzEy
Authorization: Basic dXNlcjoxMjM0
Authorization: Basic dXNlcjo2NzEy
Authorization: Basic dXNlcjE6MTIzNA==
Authorization: Basic dXNlcjE6NjcxMg==
Authorization: Basic dXNlcjI6MTIzNA==
Authorization: Basic dXNlcjI6NjcxMg==
```

## Commands

Print all combination of username/password, where usernames replace &TOKEN_0& and passwords replace &TOKEN_1&.

```sh
tokenish "Authentication: Basic &TOKEN_0&:&TOKEN_1&" -t /path/to/usernames/dir/ /path/to/passwords.txt
```

Save all combination of username/password in directory /path/output/ composed by files of 10 000 lines.

```sh
tokenish "Authentication: Basic &TOKEN_0&:&TOKEN_1&" -t /path/to/usernames/dir/ /path/to/passwords.txt -o /path/output/
```

Save all combination of username/password in directory /path/output/ composed by files of 500 lines.

```sh
tokenish "Authentication: Basic &TOKEN_0&:&TOKEN_1&" -t /path/to/usernames/dir/ /path/to/passwords.txt -o /path/output/ -om 500
```

Print all combination of username/password, where usernames replace &TOKEN_0& and passwords replace &TOKEN_1&.
Expression in tag &ENC[...]ODE& will encode in base64.

```sh
tokenish "Authentication: Basic &ENC[&TOKEN_0&:&TOKEN_1&]ODE&" -t /path/to/usernames/dir/ /path/to/passwords.txt -e base64
```

## License

tokenish is released under MIT license. See [LICENSE](https://github.com/VictorMeyer77/tokenish/blob/main/LICENSE).

## Tests

Run these commands to run tests and update coverage badge:

```sh
pip install readme-coverage-badger
coverage run -m --omit "/usr/local/*,/usr/lib/*,tests/*,venv/*" unittest discover tests/tokenish
readme-cov
```

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

Log types:

- `added` for new features.
- `changed` for changes in existing functionality.
- `fixed` for any bug fixes.
- `removed` for features removed in this release.
- `security` to invite users to upgrade in case of vulnerabilities.

## [Unreleased]

### Added
### changed
### Fixed
### removed
### security

## [0.1.0] - 2023/03/02

### Added

- Initial release


test/test_main.py
import unittest
import main
import os
from lib.result_writer import ResultWriter


class TestMain(unittest.TestCase):

    def test_main_write_filled_expressions_when_output_is_defined(self):
        token_paths = ["test/resources/input/links/1.txt",
                       "test/resources/input/usernames",
                       "test/resources/input/passwords"]
        main.main("Link: &ENC[test]ODE& &TOKEN_0& - Auth:&TOKEN_1&/&TOKEN_2&", token_paths,
                  None, ResultWriter("test/resources/output", 10000))
        output_path = "test/resources/output/part_0.txt"
        output = open(output_path, "r")
        lines = output.readlines()
        output.close()
        assert len(lines) == 40
        for line in lines:
            assert "&TOKEN" not in line
        os.remove(output_path)

    def test_main_write_filled_expressions_when_encoding_and_output_is_defined(self):
        token_paths = ["test/resources/input/links/1.txt",
                       "test/resources/input/usernames",
                       "test/resources/input/passwords"]
        main.main("Link: &ENC[test]ODE& &TOKEN_0& - Auth:&TOKEN_1&/&TOKEN_2&", token_paths,
                  "base64", ResultWriter("test/resources/output", 10000))
        output_path = "test/resources/output/part_0.txt"
        output = open(output_path, "r")
        lines = output.readlines()
        output.close()
        assert len(lines) == 40
        for line in lines:
            assert "dGVzdA==" in line
            assert "&TOKEN" not in line
        os.remove(output_path)

    def test_parse_args_return_all_parameters(self):
        assert main.parse_args(["pattern"]) == ("pattern", [], None, None)
        assert main.parse_args(["pattern", "-t", "/path/tokenA/", "/path/tokenB.txt"]) == \
               ("pattern", ["/path/tokenA/", "/path/tokenB.txt"], None, None)
        assert main.parse_args(["pattern", "-t", "/path/tokenA/", "/path/tokenB.txt", "-e", "base64"]) == \
               ("pattern", ["/path/tokenA/", "/path/tokenB.txt"], "base64", None)
        args_with_output = main.parse_args(["pattern", "-o", "/output"])
        assert args_with_output[3].directory_path == ResultWriter("/output", 10000).directory_path
        assert args_with_output[3].max_row_per_file == ResultWriter("/output", 10000).max_row_per_file
        args_with_output_and_max_rows = main.parse_args(["pattern", "-o", "/output", "-om", "50"])
        assert args_with_output_and_max_rows[3].max_row_per_file == ResultWriter("/output", 50).max_row_per_file


test/test_tokenish.py
import unittest
import tokenish
import os
from lib.result_writer import ResultWriter


class TestMain(unittest.TestCase):

    def test_main_write_filled_expressions_when_output_is_defined(self):
        token_paths = ["test/resources/input/links/1.txt",
                       "test/resources/input/usernames",
                       "test/resources/input/passwords"]
        tokenish.main("Link: &ENC[test]ODE& &TOKEN_0& - Auth:&TOKEN_1&/&TOKEN_2&", token_paths,
                      None, ResultWriter("test/resources/output", 10000))
        output_path = "test/resources/output/part_0.txt"
        output = open(output_path, "r")
        lines = output.readlines()
        output.close()
        assert len(lines) == 40
        for line in lines:
            assert "&TOKEN" not in line
        os.remove(output_path)

    def test_main_write_filled_expressions_when_encoding_and_output_is_defined(self):
        token_paths = ["test/resources/input/links/1.txt",
                       "test/resources/input/usernames",
                       "test/resources/input/passwords"]
        tokenish.main("Link: &ENC[test]ODE& &TOKEN_0& - Auth:&TOKEN_1&/&TOKEN_2&", token_paths,
                      "base64", ResultWriter("test/resources/output", 10000))
        output_path = "test/resources/output/part_0.txt"
        output = open(output_path, "r")
        lines = output.readlines()
        output.close()
        assert len(lines) == 40
        for line in lines:
            assert "dGVzdA==" in line
            assert "&TOKEN" not in line
        os.remove(output_path)

    def test_parse_args_return_all_parameters(self):
        assert tokenish.parse_args(["pattern"]) == ("pattern", [], None, None)
        assert tokenish.parse_args(["pattern", "-t", "/path/tokenA/", "/path/tokenB.txt"]) == \
               ("pattern", ["/path/tokenA/", "/path/tokenB.txt"], None, None)
        assert tokenish.parse_args(["pattern", "-t", "/path/tokenA/", "/path/tokenB.txt", "-e", "base64"]) == \
               ("pattern", ["/path/tokenA/", "/path/tokenB.txt"], "base64", None)
        args_with_output = tokenish.parse_args(["pattern", "-o", "/output"])
        assert args_with_output[3].directory_path == ResultWriter("/output", 10000).directory_path
        assert args_with_output[3].max_row_per_file == ResultWriter("/output", 10000).max_row_per_file
        args_with_output_and_max_rows = tokenish.parse_args(["pattern", "-o", "/output", "-om", "50"])
        assert args_with_output_and_max_rows[3].max_row_per_file == ResultWriter("/output", 50).max_row_per_file


test/lib/test_encoder.py
import unittest
from lib import encoder


class TestEncoder(unittest.TestCase):

    def test_apply_encoding_raises_error_when_encoding_is_unknown(self):
        with self.assertRaises(ValueError):
            encoder.apply_encoding("", "unknown")

    def test_apply_encoding_base64_returns_encoded_str(self):
        assert encoder.apply_encoding("hello world", "baSe64") == "aGVsbG8gd29ybGQ="

    def test_encode_expression_raises_error_when_tokens_are_malformed(self):
        with self.assertRaises(ValueError):
            encoder.encode_expression("&ENC[Hello", "base64")
        with self.assertRaises(ValueError):
            encoder.encode_expression("Hello]ODE&", "base64")

    def test_encode_expression_returns_encoded_target(self):
        assert encoder.encode_expression("&ENC[Hello]ODE&, World", "base64") == "SGVsbG8=, World"
        assert encoder.encode_expression("Hello, World", "base64") == "SGVsbG8sIFdvcmxk"

    def test_encode_expressions_returns_encoded_target(self):
        assert encoder.encode_expressions(["&ENC[Hello]ODE&, World", "Hello, World"], "base64") == \
               ["SGVsbG8=, World", "SGVsbG8sIFdvcmxk"]


tokenish/tokenish.py
import argparse
import sys
from tokenish.lib.result_writer import ResultWriter
from tokenish.lib.tokenizer import *
from tokenish.lib.encoder import encode_expressions


def main(pattern, token_paths, encoding, writer):
    token_list = gather_tokens(token_paths)
    filled_expressions = fill_tokens(token_list, pattern) if len(token_paths) > 0 else []
    encoded_expressions = encode_expressions(filled_expressions, encoding) if encoding is not None \
        else filled_expressions
    if writer is not None:
        [writer.write_row(expression) for expression in encoded_expressions]
    else:
        [print(expression) for expression in encoded_expressions]
    if writer is not None:
        writer.close_file()


def parse_args(args):
    parser = argparse.ArgumentParser(prog="tokenish",
                                     description="Generate rows from pattern for each token combinations",
                                     epilog="More information on GitLab")
    parser.add_argument("pattern", help="text to fill with links or usernames/passwords")
    parser.add_argument("-t", "--tokens", nargs="+", help="list of tokens file or directory path")
    parser.add_argument("-e", "--encoding", help="type of encoding to apply")
    parser.add_argument("-o", "--output-directory", help="directory where results will write, print is None")
    parser.add_argument("-om", "--max-file-rows", type=int, help="maximum number of rows per file, default 10 000")

    args = parser.parse_args(args)

    tokens = [] if args.tokens is None else args.tokens

    if args.output_directory is not None:
        max_row_per_file = 10000 if args.max_file_rows is None else args.max_file_rows
        result_writer = ResultWriter(args.output_directory, max_row_per_file)
    else:
        result_writer = None

    return args.pattern, tokens, args.encoding, result_writer


if __name__ == '__main__':
    parameters = parse_args(sys.argv[1:])
    main(pattern=parameters[0], token_paths=parameters[1], encoding=parameters[2], writer=parameters[3])
    sys.exit(0)
